### DISCUSSION TITLE
Pocketbase Auth in SvelteKit does Auth only on Server

### DISCUSSION THREAD

**Discussion by maksp-glitch** [pocketbase/pocketbase#5144] [Q&A]
Hello everybody, 

unfortunately I ran into a problem with Pocketbase in Sveltekit. 

This is my hook.server.ts

```javascript
import PocketBase from 'pocketbase';
import { env } from '$env/dynamic/public';
import { serializeNonPOJOs } from '$lib/utils';

export const handle = async ({ event, resolve }) => {
    event.locals.pb = new PocketBase(env.PUBLIC_PB_URL);
    event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');

    if (event.locals.pb.authStore.isValid) {
        event.locals.user = serializeNonPOJOs(event.locals.pb.authStore.model)
    } else {
        event.locals.user = undefined
    }

    const response = await resolve(event);
    response.headers.set('cookie', event.locals.pb.authStore.exportToCookie( {secure: false}));
    
    return response;
``` 

An this is my login action: 

```javascript
login: async ({ locals, request }) => {
        const body = Object.fromEntries(await request.formData());
        console.log(body)

        try {
            await locals.pb.collection('users').authWithPassword(body.email, body.password);

            if (!locals.pb?.authStore?.model?.verifed) {
                locals.pb.authStore.clear();
                return {
                    notVerified: true,
                    message: 'Deine E-Mail wurde noch nicht verifiziert.'
                }
            }
        } catch (err) {
            console.log(err);
            throw error(500, 'Login fehlgeschlagen');
        }

        throw redirect(303, '/account');

    }
``` 

Unfortunately the login only worked once in the browser/client. After a while i cannot login in the browser anymore. It looks like !locals.pb?.authStore?.model?.verifed references to the wrong authStore and the login method doesnt write into browser/saves in browser. But it did so first hand.

What I tried was logging the user out with a method, using another browser and restarting everything (incl. machine). Nothing worked. 

Anyone had this issue before? 

OS: Mac 
Served via Vite through npm run dev

Thank you in advance

**Comment by ganigeorgiev** [comment]
It is not clear what is the issue, nor what is the expected behavior of the above code.

`hook.server.ts` executes only on the server-side, meaning that `locals.pb` is only available in a server-side context.

If you want a mixed (both client-side and server-side) access to the PocketBase server then you'll have to create a second JS SDK instance for the client (_and if you need shared cookie access between the server and client you'll have to export it with `httpOnly:false`_).

Additionally, I've seen this several times and I don't know from where `serializeNonPOJOs` originates, but you don't need to convert the JS SDK result into POJOs because the SDK returns plain objects since v0.17+.

I've posted this several times but **if you are new to web development I don't recommend using JS SSR with PocketBase** because most likely you'll mess up the security when maintaining the auth state between the 2 backends (the SvelteKit node server and PocketBase). More ramblings on this you can read in https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7620032, https://github.com/pocketbase/pocketbase/discussions/4412#discussioncomment-8575834 and probably others.

If you need further help, please elaborate more clearly on your use case to try to understand the bigger picture and what you are trying to do.

**Reply by maksp-glitch** [reply]
Thank you for the super quick reply! 

serializeNonPOJOs comes from the huntabyte guide on pb with sveltekit. The tutorial is 1 year old and probably before v0.17. 

my problem is that if I do 
```javascript
const authData = await locals.pb.collection('users').authWithPassword(body.email, body.password);
console.log(authData)

Output: AuthData as expected with user info
``` 

and if I

```javascript
console.log(!locals.pb?.authStore?.model?.verifed)

Output: false or undefined

```
The output is odd for me since both is based on the same data, or should be. Even though it's from the same code block it gives me different outputs. 

I try to use pocketbase as my main backend for Auth, User-Management and database. It's a smaller project with a max. of 5.000 active users. 

I know that having pb running serverside comes with caveats, but i would export the user with a load function to the frontend. 
Basically in root i would create a layout.server.ts with a load function that gets the locals.pb.authStore.model like that: 

```javascript
import type { LayoutServerLoad } from './$types';

export const load = (async ({ locals }) => {
    if (locals.pb.authStore.model) {
        return { user: locals.pb.authStore.model }
    }

    return {
        user: undefined
    }
}) satisfies LayoutServerLoad;
```

**Reply by ganigeorgiev** [reply]
> The output is odd for me since both is based on the same data, or should be. Even though it's from the same code block it gives me different outputs.

I'm sorry I don't think I'll be able to help you as it is still not clear to me what actually is the problem.

Make sure that you understand where and in what order your JavaScript code is being executed.

**Reply by maksp-glitch** [reply]
thats the problem i dont get. all of the code related to PB runs on server. 

Hook in hooks.server.ts
code for /login route in page.server.ts

--> Still got the problem that the login only worked once and 
```javascript
const authData = await locals.pb.collection('users').authWithPassword(body.email, body.password);
console.log(authData)
```
gives me the correct data and 

```javascript
console.log(!locals.pb?.authStore?.model?.verifed)
```
gives me false or undefined. Even if i go with 

```javascript
//1
const authData = await locals.pb.collection('users').authWithPassword(body.email, body.password);
console.log(authData)
//2
console.log(!locals.pb?.authStore?.model?.verifed)
```

i get authData correctly printed with expected data from DB and //2 prints undefined or false.

**Reply by ganigeorgiev** [reply]
> gives me false or undefined. 

The above negated expression will never output `undefined`. Either you are not reporting it properly or the code you've shared differs from your actual one.

If you need further help, feel free to create a **minimal** reproducible repo illustrating the problem and I'll try to have a look at it but until then I don't think anyone will be able to assist you.

**Reply by maksp-glitch** [reply]
I'm 1000% embarrased... i wrote verifed instead of verified
Sorry for stealing you time for this man!
